libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp

```diff
@@ -0,0 +1,441 @@
+
+        case 6:
+            // Unsigned int base 10
+            _temp_state = strtoul(_term, nullptr, 10);
+            break;
+
+        case 7:
+            // Unsigned int base 10
+            _temp_err_code = strtoul(_term, nullptr, 10);
+            // Sentence only declared valid when we have the expected number of terms
+            _sentence_valid = true;
+            break;
+
+        default:
+            // If we get any other term number something has gone wrong
+            _sentence_valid = false;
+            break;
+    }
+}
+
+// Check for failsafes
+AP_BattMonitor::BatteryFailsafe AP_Generator_IE_FuelCell::update_failsafes() const
+{
+    if ((_state.failsafe & _crit_fs_bitmask) != 0) {
+        return  AP_BattMonitor::BatteryFailsafe::BatteryFailsafe_Critical;
+    }
+
+    if ((_state.failsafe & _low_fs_bitmask) != 0) {
+        return  AP_BattMonitor::BatteryFailsafe::BatteryFailsafe_Low;
+    }
+
+    return AP_BattMonitor::BatteryFailsafe::BatteryFailsafe_None;
+}
+
+// Lookup table for IE 650/800 W fuel cell failsafes.
+const AP_Generator_IE_FuelCell::Lookup_FS_IE6 AP_Generator_IE_FuelCell::lookup_fs_ie6[] = {
+    { ERR_CODE_IE6::STACK_OT1, (1 << 0)},
+    { ERR_CODE_IE6::STACK_OT2, (1 << 1)},
+    { ERR_CODE_IE6::BAT_UV, (1 << 2)},
+    { ERR_CODE_IE6::BAT_OT, (1 << 3)},
+    { ERR_CODE_IE6::NO_FAN, (1 << 4)},
+    { ERR_CODE_IE6::FAN_OVERRUN, (1 << 5)},
+    { ERR_CODE_IE6::STACK_OT1_2, (1 << 6)},
+    { ERR_CODE_IE6::STACK_OT2_2, (1 << 7)},
+    { ERR_CODE_IE6::BAT_UV_2, (1 << 8)},
+    { ERR_CODE_IE6::BAT_OT_2, (1 << 9)},
+    { ERR_CODE_IE6::START_TIMEOUT, (1 << 10)},
+    { ERR_CODE_IE6::STOP_TIMEOUT, (1 << 11)},
+    { ERR_CODE_IE6::START_UNDER_PRESSURE, (1 << 12)},
+    { ERR_CODE_IE6::TANK_UNDER_PRESSURE, (1 << 13)},
+    { ERR_CODE_IE6::TANK_LOW_PRESSURE, (1 << 14)},
+};
+
+// Lookup table for IE 2.4 kW fuel cell failsafes.
+const AP_Generator_IE_FuelCell::Lookup_FS_IE24 AP_Generator_IE_FuelCell::lookup_fs_ie24[] = {
+    { ERR_CODE_IE24::REDUCED_POWER, (1 << 0)},  // The unit is capible of outputting 4kW this warning is when the fuel cell has limited power output to 2.4kW
+    { ERR_CODE_IE24::SPM_LOST, (1 << 1)},
+    { ERR_CODE_IE24::PRESSURE_LOW, (1 << 2)},
+    { ERR_CODE_IE24::BATTERY_LOW, (1 << 3)},
+    { ERR_CODE_IE24::PRESSURE_ALERT, (1 << 4)},
+    { ERR_CODE_IE24::START_DENIED, (1 << 5)},
+    { ERR_CODE_IE24::SYSTEM_CRITICAL, (1 << 6)},
+    { ERR_CODE_IE24::PRESSURE_CRITICAL, (1 << 7)},
+    { ERR_CODE_IE24::BATTERY_CRITICAL, (1 << 8)},
+};
+
+// Convert error code from fuel cell to AP failsafe bitmask
+void AP_Generator_IE_FuelCell::set_failsafe_mask(uint32_t err_in)
+{
+    // If fuel cell state has changed send gcs message
+    if (_state.err_code != _state.last_err_code) {
+        if (_type == AP_Generator::GenType::IE_650_800) {
+            // Convert the error code to the failsafe bitmask for IE_650_800
+            for (const struct Lookup_FS_IE6 entry : lookup_fs_ie6) {
```

Comment:
```suggestion
            for (const auto& entry : lookup_fs_ie6) {
```

Similarly below

Reply:
Why do we want to auto here if we can easily identify the type?